### PR TITLE
Download cli from server

### PR DESCRIPTION
**Pull request by adamcarlton**

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -279,7 +287,7 @@ func init() {
 // creates our Redis client for communication
 // creates an auth client based on the environment variables and provider
 func main() {
-	if err := makeRedisClient(); err != nil {
+	if err := makeRedisClient(os.Getenv("REDIS_URL"), os.Getenv("REDIS_PASSWORD")); err != nil {
```
Comment (emmanuel) [inline cmd/server/main.go]: Do some error checking (error on blank) on `os.Getenv("REDIS_URL")` and `os.Getenv("REDIS_PASSWORD")`, like you do for `CLIENT_ID`, `CLIENT_SECRET` and `REDIRECT_URL` below.

File: cmd/server/main.go [hunk 2]
```diff
@@ -288,8 +296,23 @@ func main() {
...
+		log.Fatal("CLIENT_SECRET not set!")
+	}
+	if os.Getenv("REDIRECT_URL") == "" {
+		log.Fatal("REDIRECT_URL not set!")
+	}
 	authClient := newAuthClient(os.Getenv("CLIENT_ID"), os.Getenv("CLIENT_SECRET"), os.Getenv("REDIRECT_URL"), provider)
-	if err := http.ListenAndServe(listenPort, getMux(authClient)); err != nil {
-		log.Fatal(err)
+	var newMux *http.ServeMux
+	if os.Getenv("DOWNLOAD_DIR") == "" {
```
Comment (emmanuel) [inline cmd/server/main.go]: Decide on the value of `downloadDir` beforehand, then make one call to `getMux`.

Comment (emmanuel) [reply cmd/server/main.go]: Also, rename `newMux` to just `mux`.

Comment (adamcarlton) [reply cmd/server/main.go]: done and done

File: cmd/server/main.go (hunk 1 above)
Comment (adamcarlton) [reply cmd/server/main.go]: done-zo!

File: Dockerfile [hunk 3]
```diff
@@ -1,3 +1,6 @@
 FROM quay.io/nordstrom/baseimage-ubuntu:16.04
 COPY build/kubelogin /kubelogin
+COPY build/download/windows/kubelogin-cli-windows.zip /download/windows/
```
Comment (emmanuel) [inline Dockerfile]: The `COPY` command in Dockerfiles supports copying whole directories: https://docs.docker.com/engine/reference/builder/#copy

Therefore, please collapse these three lines to:

```
COPY build/download /download
```

File: cmd/server/main.go [hunk 4]
```diff
@@ -288,8 +302,25 @@ func main() {
...
+	if os.Getenv("CLIENT_SECRET") == "" {
+		log.Fatal("CLIENT_SECRET not set!")
+	}
+	if os.Getenv("REDIRECT_URL") == "" {
+		log.Fatal("REDIRECT_URL not set!")
+	}
 	authClient := newAuthClient(os.Getenv("CLIENT_ID"), os.Getenv("CLIENT_SECRET"), os.Getenv("REDIRECT_URL"), provider)
-	if err := http.ListenAndServe(listenPort, getMux(authClient)); err != nil {
-		log.Fatal(err)
+	var mux *http.ServeMux
```
Comment (emmanuel) [inline cmd/server/main.go]: No need to declare `mux` here and then assign to it down on line 322. Just use `:=` to declare and assign in one step. Delete line 315 and replace line 322 with:
```
mux := getMux(authClient, downloadDir)
```

File: cmd/server/main.go [hunk 5]
```diff
@@ -288,8 +302,25 @@ func main() {
...
+	var downloadDir string
```
Comment (emmanuel) [inline cmd/server/main.go]: Better style here would be to declare and assign in one step (like my comment above). Then check if blank and assign default value if need be. Like this:
```
downloadDir := os.Getenv("DOWNLOAD_DIR")
if downloadDir == "" {
    downloadDir = "/download"
}
```

**Pull request merged**